ktor-network/common/test/io/ktor/network/sockets/tests/UdpBroadcastTest.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets.tests
+
+import io.ktor.network.selector.*
+import io.ktor.network.sockets.*
+import io.ktor.test.dispatcher.*
+import io.ktor.util.*
+import io.ktor.util.network.*
+import io.ktor.utils.io.core.*
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+import kotlin.test.*
+
+class UdpBroadcastTest : CoroutineScope {
+    private val testJob = Job()
+    private val selector = SelectorManager(Dispatchers.Default + testJob)
+
+    override val coroutineContext: CoroutineContext
+        get() = testJob
+
+    @AfterTest
+    fun tearDown() {
+        testJob.cancel()
+        selector.close()
+    }
+
+    @Test
+    fun testBroadcastSuccessful() = testSuspend {
+        if (!PlatformUtils.IS_JVM) return@testSuspend
+        withTimeout(15000) {
+            val serverSocket = CompletableDeferred<BoundDatagramSocket>()
+            val server = launch {
+                aSocket(selector)
+                    .udp()
+                    .bind(NetworkAddress("0.0.0.0", 56700))
+                    .use { socket ->
+                        serverSocket.complete(socket)
+                        val received = socket.receive()
+                        assertEquals("0123456789", received.packet.readText())
+                    }
+            }
+
+            serverSocket.await()
+
+            aSocket(selector)
+                .udp()
+                .bind {
+                    broadcast = true
+                }
+                .use { socket ->
+                    socket.send(
+                        Datagram(
+                            packet = buildPacket { writeText("0123456789") },
+                            address = NetworkAddress("255.255.255.255", 56700)
+                        )
+                    )
+                }
+
+            server.join()
+        }
+    }
+
+    // TODO: this test does not catch the permission denied exception
```

Comment:
This test does not seem to catch the SocketException.
```
> Task :ktor-network:jvmTest
Exception in thread "DefaultDispatcher-worker-1 @coroutine#3" java.net.SocketException: Permission denied
	at java.base/sun.nio.ch.DatagramChannelImpl.send0(Native Method)
	at java.base/sun.nio.ch.DatagramChannelImpl.sendFromNativeBuffer(DatagramChannelImpl.java:584)
	at java.base/sun.nio.ch.DatagramChannelImpl.send(DatagramChannelImpl.java:546)
	at java.base/sun.nio.ch.DatagramChannelImpl.send(DatagramChannelImpl.java:529)
	at io.ktor.network.sockets.DatagramSocketImpl.sendImpl(DatagramSocketImpl.kt:98)
	at io.ktor.network.sockets.DatagramSocketImpl$sender$1.invokeSuspend(DatagramSocketImpl.kt:35)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

Expected an exception to be thrown, but was completed successfully.
java.lang.AssertionError: Expected an exception to be thrown, but was completed successfully.
	at org.junit.Assert.fail(Assert.java:88)
	at kotlin.test.junit.JUnitAsserter.fail(JUnitSupport.kt:56)
	at kotlin.test.AssertionsKt__AssertionsKt.checkResultIsFailure(Assertions.kt:153)
	at kotlin.test.AssertionsKt.checkResultIsFailure(Unknown Source)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1$1.invokeSuspend(UdpBroadcastTest.kt:71)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1$1.invoke(UdpBroadcastTest.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturnIgnoreTimeout(Undispatched.kt:102)
	at kotlinx.coroutines.TimeoutKt.setupTimeout(Timeout.kt:110)
	at kotlinx.coroutines.TimeoutKt.withTimeout(Timeout.kt:32)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1.invokeSuspend(UdpBroadcastTest.kt:70)
	(Coroutine boundary)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1.invokeSuspend(UdpBroadcastTest.kt:70)
	(Coroutine creation stacktrace)
	at kotlin.coroutines.intrinsics.IntrinsicsKt__IntrinsicsJvmKt.createCoroutineUnintercepted(IntrinsicsJvm.kt:122)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:26)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at io.ktor.test.dispatcher.TestJvmKt.testSuspend(TestJvm.kt:16)
	at io.ktor.test.dispatcher.TestJvmKt.testSuspend$default(TestJvm.kt)
	at io.ktor.network.sockets.tests.UdpBroadcastTest.testBroadcastFails(UdpBroadcastTest.kt:68)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.AssertionError: Expected an exception to be thrown, but was completed successfully.
	at org.junit.Assert.fail(Assert.java:88)
	at kotlin.test.junit.JUnitAsserter.fail(JUnitSupport.kt:56)
	at kotlin.test.AssertionsKt__AssertionsKt.checkResultIsFailure(Assertions.kt:153)
	at kotlin.test.AssertionsKt.checkResultIsFailure(Unknown Source)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1$1.invokeSuspend(UdpBroadcastTest.kt:71)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1$1.invoke(UdpBroadcastTest.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturnIgnoreTimeout(Undispatched.kt:102)
	at kotlinx.coroutines.TimeoutKt.setupTimeout(Timeout.kt:110)
	at kotlinx.coroutines.TimeoutKt.withTimeout(Timeout.kt:32)
	at io.ktor.network.sockets.tests.UdpBroadcastTest$testBroadcastFails$1.invokeSuspend(UdpBroadcastTest.kt:70)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:81)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:56)
	... 52 more

io.ktor.network.sockets.tests.UdpBroadcastTest[jvm] > testBroadcastFails[jvm] FAILED
    java.lang.AssertionError at UdpBroadcastTest.kt:68
        Caused by: java.lang.AssertionError at UdpBroadcastTest.kt:68
1 test completed, 1 failed
```

Reply:
Issue reported on YouTrack: https://youtrack.jetbrains.com/issue/KTOR-1166

Reply:
The reason is that datagrams are being sent through a channel so a failure state will be propagated asynchronously. It's not easy to catch it in the test

Reply:
Should be fixed with https://github.com/ktorio/ktor/pull/2164